### DISCUSSION TITLE
Access a variable or function in main.py from another file.

### DISCUSSION THREAD

**Discussion by ifamirhasan** [fastapi/fastapi#6469] [Questions] [answered]
### First Check

- [x] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
no code :(
```

### Description

I want to get access to a variable or function which is in main.py from another file (a file which has routers).
for example I create a variable (request rate limiter) in main.py and pass it to fastapi app. then I want to use it as decorator when defining routes but my routes are not in main.py so I have to pass to router file to use it while creating routes.
I used this example (rate limiter) to just explain my problem and there might be better ways for setting up this and may no need for passing variable like this.
Thank you in advance

### Operating System

Linux

### FastAPI Version

0.65.2

### Python Version

3.8.6

**Comment by jgould22** [comment]
If you place `__init__.py` files in the correct folders you can simply import your function with something like `from <project_name>.main import rate_limit_decorator`

**Comment by diegocgaona** [comment]
The really good FastAPI documentation have a nice tutorial which could help:
https://fastapi.tiangolo.com/tutorial/bigger-applications/

Maybe would be nice to see dependencies too:
https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/

I would make a dependencies.py file and import it from where you need.

**Comment by ifamirhasan** [comment]
@diegocgaona 
Thanks for your suggestion but I read these documents, in my case they don't meet my needs.
because I need to pass a variable before initializing my routes. and I couldn't find anything about it.

 @jgould22 
I don't have __init__.py in the root of my project. I thought I should use when I want to create new module so can you explain it a little more? I'm new at python I might understand import system wrong.
But recently I thought about a way to pass my variable to another file but I'm not sure if it is good way or not?
for example I can pass my variable (rateLimiter) to a config file then import myfile. So I can  get the variable from config to initialize my routes.
Thanks for your help.

**Comment by diegocgaona** [comment]
You can do it, create an dependencies.py (or any name you want), import it in your main (imports are made on the top of the file, and you can follow what @jgould22 said or search about "python import variable from another file" on Google) and use it in the line after the import.

I will try to make a generic example: (basic from the first link I send, but mixed with first steps)


The structure of files would be like that: (__init__.py is a blank file)

```
├── app
│   ├── __init__.py
│   ├── main.py
│   └── dependencies.py
```

Your main file:
```
from fastapi import FastAPI
# I think the standard name in Python would be rate_limiter, if it's a function and RateLimiter if it a class
from ..dependencies import rateLimiter  

app = FastAPI()


# Use it where you need
@app.get("/")
@rateLimiter  # Your decorator
async def root():
    return {"message": "Hello World"}
```

If I understood correctly, it is what you want, but without code and directory structure, is hard to say and I'm not an expert.

**Comment by tonyhart7** [comment]
> You can do it, create an dependencies.py (or any name you want), import it in your main (imports are made on the top of the file, and you can follow what @jgould22 said or search about "python import variable from another file" on Google) and use it in the line after the import.
> 
> I will try to make a generic example: (basic from the first link I send, but mixed with first steps)
> 
> The structure of files would be like that: (**init**.py is a blank file)
> 
> ```
> ├── app
> │   ├── __init__.py
> │   ├── main.py
> │   └── dependencies.py
> ```
> 
> Your main file:
> 
> ```
> from fastapi import FastAPI
> # I think the standard name in Python would be rate_limiter, if it's a function and RateLimiter if it a class
> from ..dependencies import rateLimiter  
> 
> app = FastAPI()
> 
> 
> # Use it where you need
> @app.get("/")
> @rateLimiter  # Your decorator
> async def root():
>     return {"message": "Hello World"}
> ```
> 
> If I understood correctly, it is what you want, but without code and directory structure, is hard to say and I'm not an expert.

this works Iam import my subfolder on main.py 

but I cant access app module on subfolder why is that ?? 

my structure project 
-app
-main.py
-models.py
---router
---init
---post.py

when on post.py 
I cant access models under app folder 
from app import models, schemas

why is that ?

**Comment by jgould22** [comment]
I cant really understand your folder structure diagram but  you can look at the FastAPI cookie cutter template to see how some things are done https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app that might help you along a bit

**Comment by tonyhart7** [comment]
> I cant really understand your folder structure diagram but you can look at the FastAPI cookie cutter template to see how some things are done https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app that might help you along a bit

Iam following course right now to learn fastapi and python 

so project generate is a lil bit overwheilming for me because its first time I created backend service

**Comment by ifamirhasan** [comment, marked as answer]
> You can do it, create an dependencies.py (or any name you want), import it in your main (imports are made on the top of the file, and you can follow what @jgould22 said or search about "python import variable from another file" on Google) and use it in the line after the import.
> 
> I will try to make a generic example: (basic from the first link I send, but mixed with first steps)
> 
> The structure of files would be like that: (**init**.py is a blank file)
> 
> ```
> ├── app
> │   ├── __init__.py
> │   ├── main.py
> │   └── dependencies.py
> ```
> 
> Your main file:
> 
> ```
> from fastapi import FastAPI
> # I think the standard name in Python would be rate_limiter, if it's a function and RateLimiter if it a class
> from ..dependencies import rateLimiter  
> 
> app = FastAPI()
> 
> 
> # Use it where you need
> @app.get("/")
> @rateLimiter  # Your decorator
> async def root():
>     return {"message": "Hello World"}
> ```
> 
> If I understood correctly, it is what you want, but without code and directory structure, is hard to say and I'm not an expert.

Thanks for your answer and help.
I understand what you are saying but the rate limiter is just an example. In my case it is different and I have to do few things with that variable before app starts and I can't create it for each request. but fortunately I figured it out, as I said in my last comment I have to create the variable in a file (config.py) and I can import it every where and use it globally.
Again thanks for your help.
